feat: add terraform solution for autoscaling-model-server 0.1 (#4)
canonical/autoscaling-model-server: terraform/README.md

# Autoscaling model server Terraform solution

This is a Terraform module facilitating the deployment of the Autoscaling model server solution, using the [Terraform juju provider](https://github.com/juju/terraform-provider-juju/). For more information, refer to the provider [documentation](https://registry.terraform.io/providers/juju/juju/latest/docs). 

## API

### Inputs
The solution module offers the following configurable inputs:

| Name | Type | Description | Required |
| - | - | - | - |
| `<charm_name>_revision`| number | For each charm of the solution, the revision of the charm to deploy | False |

> | `cos_configuration`| bool | Boolean value that enables COS configuration | False |

Reviewer: What about creating a model? In other solutions supported by the team, we always used the `create_model` bool which would enable the creation of a model directly by solution module. We can investigate the option of skipping it here but bare in mind that then the autoscaling solution module will always have to be part of a higher level module, since it will require a [juju_model.name](https://registry.terraform.io/providers/juju/juju/latest/docs/resources/model) resource to be passed.

Author: See https://github.com/canonical/autoscaling-model-server/pull/4#discussion_r1937240211

| `existing_grafana_agent_name`| string | Name of an existing grafana-agent-k8s deployment | False |
| `grafana_agent_k8s_size`| string | Grafana agent database storage size | False |

> | `module_name`| string | Name of the model that the charm is deployed on | True |

Reviewer: ```suggestion
| `model`| string | Name of the Juju model for the deployment | True |
```
1. We had decided to use `model` instead of `model_name` as per CC006 requirements. Note that charms have `model_name` because they were written before the change in the spec.
2. Updated the wording due to it being a bundle and not a charm

Author: I removed this because the `model` is no longer an input.

### Outputs
Upon applied, the solution module exports the following outputs:

| Name | Description |
| - | - |
| `grafana_agent_k8s`| Map containing the `app_name`, `provides` and `requires` endpoints of the grafana-agent-k8s charm used |

## Usage

> This solution module is intended to be used either on its own or as part of a higher-level module. 

Reviewer: I think that this is not 100%. Since the module doesn't create a juju model, it won't be able to be used on its own.

Author: What's the best suggestion here? Should we create the juju model?

Author: Based on an offline conversation with @orfeas-k, we decided that it is better to create a Juju model if the `create_model` option is set to `True`; otherwise skip model creation and re-use the model defined in a higher level module.

This is the standard followed in other solutions from the team, as well is recommended to avoid:

```
juju add-model test
terraform apply -var model="test"
```

### COS configuration

#### Enable COS configuration
The `cos_configuration` input enables the solution to configure Charmed Kubeflow to integrate with COS. This is done by deploying a `grafana-agent-k8s` charm and adding all the required relations.
```
terraform apply -var cos_configuration=true

> ```

Reviewer: I don't think this will work as is because the solution module needs a model.

Author: Agreed, see https://github.com/canonical/autoscaling-model-server/pull/4#discussion_r1937240211

#### Use an existing grafana-agent-k8s

> If there is already an instance of the grafana-agent-k8s charm in the `kubeflow` model, then it can be used instead of deploying a new one. This is achieved with the use of `existing_grafana_agent_name` input. By default, its value is `null`.

Reviewer: **suggestion**: Do we need to keep model kubeflow?

Author: Leftover from another README, will remove!

```
terraform apply -var cos_configuration=true -var existing_grafana_agent_name="dummy-grafana-agent"

> ```

Reviewer: same as previously, this needs a model

Author: See https://github.com/canonical/autoscaling-model-server/pull/4#discussion_r1937240211
